src/core/shape_inference/include/embedding_shape_infer_utils.hpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+
+#include "openvino/core/node.hpp"
+
+namespace ov {
+namespace op {
+namespace util {
+namespace embedding {
+/**
+ * \brief Return a copy of the source shape with the first dimension replaced by provided dim
+ *
+ * \tparam TShape        Shape type
+ * \tparam TDim          Dimension type
+ *
+ * \param op             Pointer to operator.
+ * \param shape_src      The shape to be copied
+ * \param dim            The replacement dimension to overwirite the first one
+ *
+ * \return The copy of the `shape_src` with `dim` as the first element if the rank is static,
+ *         otherwise fully dynamic shape with dynamic rank
+ */
+template <class TShape, class TDim = typename TShape::value_type>
+TShape copy_shape_and_update_dim(const ov::Node* op, const TShape& shape_src, const TDim& dim) {
+    if (shape_src.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op, shape_src.size() > 0, "EMB_TABLE can't be a scalar.");
+        auto shape_copy = shape_src;
+        shape_copy[0] = dim;
+        return shape_copy;
+    }
+    return PartialShape::dynamic();
+}
```

Comment:
```suggestion
template <class TShape>
TShape out_shape_infer(const ov::Node* op, const TShape& emb_table_shape, const TShape& batch_shape) {
    auto out_shape = emb_table_shape;
    if (emb_table_shape.rank().is_static()) {
        NODE_VALIDATION_CHECK(op, emb_table_shape.size() > 0, "EMB_TABLE can't be a scalar.");
        shape_copy[0] = batch_shape.rank().is_static() ? batch_shape[0] : Dimension::dynamic());
    }
    return out_shape;
}
```
Just one function version (maybe name it out_shape_calc)

Reply:
The second version is intended to be used for [EmbeddingSegmentsSum](https://github.com/openvinotoolkit/openvino/pull/16119) as there the dimension is taken from the input value, not from the other shape.

Reply:
In mentioned PR takes input as shape also after modifications.

Reply:
As reusage of the helper for the EmbeddingSegmentsSum can lead to redundant checks,
I've updated the code to keep the only one version.